### ISSUE TITLE
`possibly-undefined` false positive with `global` statement inside `if`

### ISSUE THREAD

**Issue by andersk** [python/mypy#15838]
**Bug Report**

The optional `possibly-undefined` error gives a false positive on this code with a `global` statement inside an `if` branch. At runtime, Python treats `global` as applying unconditionally for the entire function scope, but mypy seems to think it only applies to the branch.

**To Reproduce**

```python
# mypy: enable-error-code="possibly-undefined"

g = 1

def f(b: bool) -> None:
    if b:
        global g
        g = 2

    print(g)  # error: Name "g" may be undefined  [possibly-undefined]
```

[Playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=2097ffd998af1da20198f4a2704162b5)

**Expected Behavior**

No error.

**Actual Behavior**

```
main.py:10: error: Name "g" may be undefined  [possibly-undefined]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.4.1
- Mypy command-line flags: `--enable-error-code=possibly-undefined`
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10.12
